nn/transformer.py

```diff
@@ -0,0 +1,242 @@
+    self.dropout = dropout
+
+  def forward(self, tgt: nn.LayerRef, memory: nn.LayerRef) -> nn.LayerRef:
+    """
+    Two possible forward variants of decoder, defined by self.norm_first
+    """
+    x = tgt
+    if self.norm_first:
+      x = x + self._self_attention_block(self.norm(x, epsilon=self.norm_eps))
+      x = x + self._multi_head_attention_block(self.norm(x, epsilon=self.norm_eps), memory)
+      x = x + self._feed_forward_block(self.norm(x, epsilon=self.norm_eps))
+    else:
+      x = self.norm(x + self._self_attention_block(x), epsilon=self.norm_eps)
+      x = self.norm(x + self._multi_head_attention_block(x, memory), epsilon=self.norm_eps)
+      x = self.norm(x + self._feed_forward_block(x), epsilon=self.norm_eps)
+
+    return x
+
+  def _self_attention_block(self, x: nn.LayerRef) -> nn.LayerRef:
+    x = self.self_attn(x, x, x)
+    return nn.dropout(x, self.dropout)
+
+  def _multi_head_attention_block(self, x: nn.LayerRef, mem: nn.LayerRef) -> nn.LayerRef:
+    x = self.multihead_attn(x, mem, mem)
+    return nn.dropout(x, self.dropout)
+
+  def _feed_forward_block(self, x: nn.LayerRef) -> nn.LayerRef:
+    x = self.linear_ff(x)
+    x = self.activation(x)
+    x = nn.dropout(x, dropout=self.dropout)
+    x = self.linear_out(x)
+    x = nn.dropout(x, dropout=self.dropout)
+    return x
+
+
+class TransformerDecoder(nn.Module):
+  """
+  Defines the full Decoder of the standard transformer
+  """
+  def __init__(self, decoder_layer: Union[TransformerDecoderLayer, Any], num_layers: int,
+               norm: Callable[[nn.LayerRef], nn.LayerRef] = None, norm_eps: float = 1e-5):
+      """
+      :param decoder_layer: Decoder layer to be stacked num_layers times
+      :param num_layers: Number of layers
+      :param norm: normalization function for output layer normalization
+      :param norm_eps: Epsilon value for output layer normalization
+      """
+      super(TransformerDecoder, self).__init__()
+      import copy
+      self.layers = nn.ModuleList([copy.deepcopy(decoder_layer) for _ in range(num_layers)])
+
+      self.num_layers = num_layers
+      self.norm = norm
+      self.norm_eps = norm_eps
+
+  def forward(self, tgt: nn.LayerRef, memory: nn.LayerRef) -> nn.LayerRef:
+    """
+    Executes the decoder layer as often as in num_layers defined
+    """
+    output = tgt
+    for mod in self.layers:
+      output = mod(output, memory)
+
+    if self.norm is not None:
+      output = self.norm(output, epsilon=self.norm_eps)
+
+    return output
+
+
+class Transformer(nn.Module):
+  """
+  Standard Transformer Module
+  """
+  def __init__(self, dim_model: int = 512, num_heads: int = 8, num_encoder_layers: int = 6,
+               num_decoder_layers: int = 6, dim_ff: int = 2048, drop: float = 0.1,
+               activation: Callable[[nn.LayerRef], nn.LayerRef] = nn.relu,
+               custom_encoder: Optional[Any] = None, custom_decoder: Optional[Any] = None,
+               norm_eps: float = 1e-5) -> None:
```

Comment:
Why is there no `norm` argument here?

Reply:
This is took on from PyTorch, do you think it would make sense here? Usually the standard Transformer uses the usual layer normalization right? But I can add it if you want.

Reply:
But this is inconsistent or not? Why is it configurable for all the other modules but not here?

Reply:
yes makes sense.